File: tests/AcceptanceTests/StepDefinitions/HomePageSteps.cs
// ============================================================================
// RAJ Financial - Home Page Step Definitions
// ============================================================================
// Reqnroll step definitions for HomePage.feature
// ============================================================================

using Microsoft.Playwright;
using RajFinancial.AcceptanceTests.Hooks;
using Reqnroll;

namespace RajFinancial.AcceptanceTests.StepDefinitions;

/// <summary>
/// Step definitions for home page scenarios.
/// </summary>
[Binding]
public class HomePageSteps(ScenarioContext scenarioContext)
{
    private IPage Page => scenarioContext.GetPage();

    [Given(@"I am on the home page")]
    public async Task GivenIAmOnTheHomePage()
    {
        await Page.GotoAsync(PlaywrightHooks.BaseUrl, new PageGotoOptions
        {
            WaitUntil = WaitUntilState.NetworkIdle
        });
        // Wait for Blazor to hydrate
        await Page.WaitForTimeoutAsync(2000);
    }

    [Then(@"I should see the brand name ""(.*)""")]
    public async Task ThenIShouldSeeTheBrandName(string brandName)
    {
        var content = await Page.ContentAsync();
        Assert.Contains(brandName, content);
    }

    [Then(@"I should see the tagline ""(.*)""")]
    public async Task ThenIShouldSeeTheTagline(string tagline)
    {
        var content = await Page.ContentAsync();
        Assert.Contains(tagline, content);
    }

    [Then(@"I should see the hero section")]
    public async Task ThenIShouldSeeTheHeroSection()
    {
        var heroSection = await Page.Locator(".hero-section").CountAsync();
        Assert.True(heroSection >= 1, "Hero section should be visible");
    }


    [Then(@"I should see at least (\d+) feature cards")]
    public async Task ThenIShouldSeeAtLeastFeatureCards(int count)
    {
        var featureCards = await Page.Locator(".feature-card").CountAsync();
        Assert.True(featureCards >= count, $"Expected at least {count} feature cards, found {featureCards}");
    }

    [Then(@"I should see features describing the platform benefits")]
    public async Task ThenIShouldSeeFeaturesDescribingThePlatformBenefits()
    {
        var featureCards = await Page.Locator(".feature-card").CountAsync();
        Assert.True(featureCards >= 1, "Should have feature cards describing benefits");
    }

    [Then(@"I should see the CTA section")]
    public async Task ThenIShouldSeeTheCtaSection()
    {
        var ctaSection = await Page.Locator(".cta-section").CountAsync();
        Assert.True(ctaSection >= 1, "CTA section should be visible");
    }

    [Then(@"the CTA section should encourage users to sign up")]
    public async Task ThenTheCtaSectionShouldEncourageUsersToSignUp()
    {
        var ctaContent = await Page.Locator(".cta-section").TextContentAsync();
        Assert.NotNull(ctaContent);
        // Should have some call to action text
        Assert.True(ctaContent.Length > 10, "CTA section should have content");
    }


    [Then(@"the hero section should be visible")]
    public async Task ThenTheHeroSectionShouldBeVisible()
    {
        await ThenIShouldSeeTheHeroSection();
    }

    [Then(@"the navigation should be accessible")]
    public async Task ThenTheNavigationShouldBeAccessible()
    {
        // On mobile, navigation is accessible via hamburger menu
        var hamburger = await Page.Locator("[aria-label*='menu'], [aria-label*='Menu'], .hamburger, .menu-toggle").CountAsync();
        var nav = await Page.Locator("nav").CountAsync();
        Assert.True(hamburger >= 1 || nav >= 1, "Navigation should be accessible");
    }

    [Then(@"the feature cards should be visible")]
    public async Task ThenTheFeatureCardsShouldBeVisible()
    {
        var featureCards = await Page.Locator(".feature-card").CountAsync();
        Assert.True(featureCards >= 1, "Feature cards should be visible");
    }


    [Then(@"focus indicators should be visible when tabbing")]
    public async Task ThenFocusIndicatorsShouldBeVisibleWhenTabbing()
    {
        await Page.Keyboard.PressAsync("Tab");
        await Page.Keyboard.PressAsync("Tab");
        
        var hasFocusIndicator = await Page.EvaluateAsync<bool>(@"
            () => {
                const el = document.activeElement;
                if (!el) return false;
                const styles = window.getComputedStyle(el);
                const outline = styles.outline;
                const boxShadow = styles.boxShadow;
                return (outline && !outline.includes('none')) || (boxShadow && boxShadow !== 'none');
            }
        ");
        Assert.True(hasFocusIndicator, "Focus indicators should be visible");
    }
}

File: tests/e2e/step-definitions/home-page.steps.ts
// ============================================================================
// Home Page Step Definitions
// ============================================================================
// Steps for HomePage.feature — tests the public landing page.
// Selectors target React components with data-testid attributes.
// ============================================================================

import { Given, Then } from "@cucumber/cucumber";
import { expect } from "@playwright/test";
import { CustomWorld } from "../support/world";
import { config } from "../support/config";

Given("I am on the home page", async function (this: CustomWorld) {
  await this.page.goto(config.baseUrl, { waitUntil: "networkidle" });
  // Wait for React to render
  await this.page.waitForTimeout(1000);
});

Then(
  "I should see the brand name {string}",
  async function (this: CustomWorld, brandName: string) {
    // Check logo alt text — brand name appears on the <img> element, not as visible text
    await expect(
      this.page.getByAltText(brandName).first()
    ).toBeVisible({ timeout: 5000 });
  }
);

Then(
  "I should see the tagline {string}",
  async function (this: CustomWorld, tagline: string) {
    // Tagline text spans multiple elements inside the hero <h1>.
    // innerText respects CSS layout (block spans → newlines) whereas
    // textContent concatenates without any separator.
    const hero = this.page.locator("[data-testid='hero-section']");
    await expect(hero).toBeVisible({ timeout: 5000 });
    const heroText = await hero.innerText();
    const normalised = heroText.replace(/\s+/g, " ");
    if (!normalised.includes(tagline)) {
      throw new Error(
        `Expected hero section to contain tagline "${tagline}" but got: "${normalised.substring(0, 300)}…"`
      );
    }
  }
);

Then("I should see the hero section", async function (this: CustomWorld) {
  const hero = this.page.locator("[data-testid='hero-section']");
  await expect(hero).toBeVisible({ timeout: 5000 });
});

Then(
  "I should see at least {int} feature cards",
  async function (this: CustomWorld, count: number) {
    const cards = this.page.locator("[data-testid='feature-card']");
    const actual = await cards.count();
    expect(actual).toBeGreaterThanOrEqual(count);
  }
);

Then(
  "I should see features describing the platform benefits",
  async function (this: CustomWorld) {
    const cards = await this.page
      .locator("[data-testid='feature-card']")
      .count();
    expect(cards).toBeGreaterThanOrEqual(1);
  }
);

Then("I should see the CTA section", async function (this: CustomWorld) {
  const cta = this.page.locator("[data-testid='cta-section']");
  await expect(cta).toBeVisible({ timeout: 5000 });
});

Then(
  "the CTA section should encourage users to sign up",
  async function (this: CustomWorld) {
    const ctaText = await this.page
      .locator("[data-testid='cta-section']")
      .textContent();
    expect(ctaText).not.toBeNull();
    expect(ctaText!.length).toBeGreaterThan(10);
  }
);

Then(
  "the hero section should be visible",
  async function (this: CustomWorld) {
    const hero = this.page.locator("[data-testid='hero-section']");
    await expect(hero).toBeVisible({ timeout: 5000 });
  }
);

Then(
  "the navigation should be accessible",
  async function (this: CustomWorld) {
    const hamburger = await this.page
      .locator(
        "[aria-label*='menu' i], [aria-label*='Menu'], [data-testid='mobile-menu-toggle']"
      )
      .count();
    const nav = await this.page.locator("nav").count();
    expect(hamburger >= 1 || nav >= 1).toBe(true);
  }
);

Then(
  "the feature cards should be visible",
  async function (this: CustomWorld) {
    const cards = await this.page
      .locator("[data-testid='feature-card']")
      .count();
    expect(cards).toBeGreaterThanOrEqual(1);
  }
);
